src/ray/gcs/gcs_server/gcs_resource_manager.cc

```diff
@@ -276,6 +283,10 @@ void GcsResourceManager::OnNodeAdd(const rpc::GcsNodeInfo &node) {
     RAY_LOG(WARNING) << "The registered node " << NodeID::FromBinary(node.node_id())
                      << " doesn't set the total resources.";
   }
+  rpc::ResourcesData data;
+  data.set_node_id(node.node_id());
+  data.set_node_manager_address(node.node_manager_address());
+  node_resource_usages_.emplace(NodeID::FromBinary(node.node_id()), std::move(data));
```

Comment:
Why init resource usages?

Reply:
You never know when will RPC callback be called. Add it here and remove it when node deleted is for safety. Otherwise, you might have issues when the node is deleted and then rpc is returned.

I believe the current code path for resources has the same issue. I'm not fixing it there in this PR.
Besides, in the future, I'll remove node manager address from the protobuf which is not necessary as well.

Reply:
This is added in `OnNodeAdded`, how does it affect node deletion?
besides we check if the node is still in `node_resource_usages_ ` before accessing it. like here: https://github.com/ray-project/ray/blob/master/src/ray/gcs/gcs_server/gcs_resource_manager.cc#L227

Reply:
In the code, we delete the data  https://github.com/ray-project/ray/blob/6560a2bf5df02ba7a975c249698f90bd850c2390/src/ray/gcs/gcs_server/gcs_resource_manager.cc#L293

Super wired to me we don't add it and we only delete it.

Reply:
It's added implicitly.